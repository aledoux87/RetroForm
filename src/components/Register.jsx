import React, { useState } from 'react';
import { string, func } from 'prop-types';
import Field from './Field';
import UserService from '../services/userService';
import { useHistory } from 'react-router-dom';

// class Reg extends React.Component {
//   constructor(props) {
//     super(props);

//     this.state = {
//       firstName: '',
//       lastName: ''
//     };
//   }

//   componentDidMount() {
//     this.setState({ lastName: 'plop' });
//   }

//   componentDidUpdate(currentState, prevState) {
//     if (currentState !== prevState) {
//       // TODO: ...
//     }
//   }

//   render() {
//     return <div></div>;
//   }
// }

function Register(props) {
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [yearOld, setYearOld] = useState(0);
  const [gender, setGender] = useState('unknown');

  React.useEffect(() => {
    // TODO: ...
  }, [firstName]);

  const history = useHistory();

  const submitForm = event => {
    event.preventDefault();
    UserService.registerUser({ firstName, lastName, yearOld, gender })
      .then(result => {
        props.setcurrentUser(result);
        history.push('/recap');
      })
      .catch(err => alert(err));
  };

  return (
    <form className="retro-form" onSubmit={submitForm}>
      <h2>{props.title}</h2>
      {props.subTitle && <p>{props.subTitle}</p>}

      <Field
        label="What is your first name ?"
        id="firstName"
        placeholder="John"
        defaultValue={props.firstName}
        onChange={e => setFirstName(e.target.value)}
      />

      <Field
        label="What is your last name ?"
        id="lastName"
        placeholder="Doe"
        defaultValue={props.lastName}
        onChange={e => setLastName(e.target.value)}
      />

      <Field
        label="How old are you ?"
        id="yearOld"
        type="number"
        placeholder="42"
        onChange={e => setYearOld(e.target.value)}
      />

      <div className="retro-field">
        <label htmlFor="gender">What kind of human are you ?</label>
        <select
          className="retro-select"
          id="gender"
          name="gender"
          defaultValue={props.gender}
          onChange={e => setGender(e.target.value)}>
          <option value="unknown" disabled>
            Unknown
          </option>
          <option value="gentleman">Gentleman</option>
          <option value="lady">Lady</option>
        </select>
      </div>

      <button className="retro-button" type="submit">
        SEND
      </button>
    </form>
  );
}

Register.propTypes = {
  title: string.isRequired,
  subTitle: string,
  setcurrentUser: func
};

export default Register;
