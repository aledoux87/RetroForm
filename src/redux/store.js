import { combineReducers, createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import user from './user';

const rootReducer = combineReducers({
  user
});

export const makeStore = (state = {}) => {
  // prettier-ignore
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  return createStore(rootReducer, state, composeEnhancers(applyMiddleware(thunk)));
};

const store = makeStore();

export default store;
