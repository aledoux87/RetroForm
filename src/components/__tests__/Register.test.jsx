import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { makeStore } from '../../redux/store';
import Register from '../Register';
import UserService from '../../services/userService';

jest.mock('../../services/userService');

describe('<Register />', () => {
  // const { location } = window;
  const store = makeStore();

  // beforeAll(() => {
  //   delete window.location;
  //   window.location = {
  //     href: '',
  //   };
  // });

  // afterAll(() => {
  //   window.location = location;
  // });

  it('Should Component Renders With Default Behaviour', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Register title="LoremIpsum" />
        </Router>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('Should Component Renders With Props', () => {
    const { queryByText } = render(
      <Provider store={store}>
        <Router>
          <Register title="LoremIpsum" subTitle="Lorem ipsum dolor sit amet" />
        </Router>
      </Provider>
    );
    expect(queryByText('Lorem ipsum dolor sit amet')).toBeInTheDocument();
  });

  it("Should 'onSubmit' Function Has Been Called", async () => {
    UserService.registerUser.mockResolvedValue({
      firstName: 'John',
      lastName: 'DOE',
      birthYear: 1993,
      gender: 'Boy'
    });

    const { container } = render(
      <Provider store={store}>
        <Router>
          <Register title="LoremIpsum" />
        </Router>
      </Provider>
    );

    const firstNameInput = container.querySelector('#firstName');
    const lastNameInput = container.querySelector('#lastName');
    const yearOldInput = container.querySelector('#yearOld');
    const genderSelect = container.querySelector('#gender');

    fireEvent.change(firstNameInput, { target: { value: 'John' } });
    fireEvent.change(lastNameInput, { target: { value: 'Doe' } });
    fireEvent.change(yearOldInput, { target: { value: '42' } });
    fireEvent.change(genderSelect, { target: { value: 'gentleman' } });

    const sendButton = screen.getByText('SEND');

    fireEvent.click(sendButton);

    await waitFor(() => {
      expect(window.location.href).toEqual('http://localhost/listing');
    });
  });
});
