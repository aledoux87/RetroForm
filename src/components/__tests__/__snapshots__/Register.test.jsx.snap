// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> Should Component Renders With Default Behaviour 1`] = `
<div>
  <form
    class="retro-form"
  >
    <h2>
      LoremIpsum
    </h2>
    <div
      class="retro-field"
    >
      <label
        for="firstName"
      >
        What is your first name ?
      </label>
      <input
        class="retro-input"
        id="firstName"
        name="firstName"
        placeholder="John"
        type="text"
        value=""
      />
    </div>
    <div
      class="retro-field"
    >
      <label
        for="lastName"
      >
        What is your last name ?
      </label>
      <input
        class="retro-input"
        id="lastName"
        name="lastName"
        placeholder="Doe"
        type="text"
        value=""
      />
    </div>
    <div
      class="retro-field"
    >
      <label
        for="yearOld"
      >
        How old are you ?
      </label>
      <input
        class="retro-input"
        id="yearOld"
        name="yearOld"
        placeholder="42"
        type="number"
        value=""
      />
    </div>
    <div
      class="retro-field"
    >
      <label
        for="gender"
      >
        What kind of human are you ?
      </label>
      <select
        class="retro-select"
        id="gender"
        name="gender"
      >
        <option
          disabled=""
          selected=""
          value="unknown"
        >
          Unknown
        </option>
        <option
          value="gentleman"
        >
          Gentleman
        </option>
        <option
          value="lady"
        >
          Lady
        </option>
      </select>
    </div>
    <button
      class="retro-button"
      type="submit"
    >
      SEND
    </button>
  </form>
</div>
`;